alicloud/resource_alicloud_cdn_domain_test.go

```diff
@@ -112,6 +112,36 @@ func TestAccAlicloudCdnDomain_basic(t *testing.T) {
 	})
 }
 
+func TestAccAlicloudCdnDomain_https(t *testing.T) {
+	var v cdn.DomainDetail
+	rand := acctest.RandInt()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+
+		// module name
+		IDRefreshName: "alicloud_cdn_domain.domain",
+
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCdnDomainDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccCdnDomainHttpsConfig(rand),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCdnDomainExists("alicloud_cdn_domain.domain", &v),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "domain_name", fmt.Sprintf("tf-testacc%d.xiaozhu.com", rand)),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.#", "1"),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.server_certificate_status", "on"),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.server_certificate", strings.Replace(testServerCertificate, `\n`, "\n", -1)),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.private_key", strings.Replace(testPrivateKey, `\n`, "\n", -1)),
```

Comment:
I think the hash code should not occur in testcase because of it always changes. If you want to check `server_certificate_status `  and other config, you can consider to modify https_config type to TypeList.

Reply:
I saw this pattern in the AWS provider, ex https://github.com/terraform-providers/terraform-provider-aws/blob/5a062fcdc5b24d82017aa15bc561fac1f152a2c6/aws/resource_aws_elb_test.go#L122. I agree it's not ideal to include the hash code in the test, but it seems like we shouldn't need to change the provider structure just to avoid dealing with hash codes in the state. The other params of `alicloud_cdn_domain` all use `schema.TypeSet`, ex `auth_config`, `refer_config`, `page_404_config`, and `parameter_filter_config` so I wanted to be consistent. Is there another helper I can use to write assertions that work on `schema.TypeSet` values, or is it not possible to test `schema.TypeSet` types?

Reply:
If you can ensure same config has same hash code, the testcase can be kept. Can you run the testcase and upload the result when next submitting?

Reply:
The hash shouldn't change unless we change the values contained in the config. Here is the output of the test:
```
=== RUN   TestAccAlicloudCdnDomain_https
2019/01/04 11:32:07 [WARN] Test: Executing step 0
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeValidate
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [WARN] Test: Step plan: DIFF:

CREATE: alicloud_cdn_domain.domain
  cdn_type:                                                "" => "web"
  certificate_config.#:                                    "0" => "1"
  certificate_config.3358202545.private_key:               "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.3358202545.server_certificate:        "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.3358202545.server_certificate_status: "" => "on"
  domain_name:                                             "" => "tf-testacc6068554464636384429.xiaozhu.com"
  scope:                                                   "" => "overseas"
  source_port:                                             "" => "80"
  source_type:                                             "" => "oss"
  sources.#:                                               "0" => "1"
  sources.21942838:                                        "" => "terraformtest.aliyuncs.com"

STATE:

<no state>
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeApply
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:48 [WARN] Domain id "tf-testacc6068554464636384429.xiaozhu.com"
2019/01/04 11:38:48 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:38:48 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:48 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:48 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [WARN] Test: Running ID-only refresh check on tf-testacc6068554464636384429.xiaozhu.com
2019/01/04 11:38:51 [WARN] Test: Executing destroy step
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypeValidate
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:55 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:55 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:55 [WARN] Test: Step plan: DIFF:

DESTROY: alicloud_cdn_domain.domain

STATE:

alicloud_cdn_domain.domain:
  ID = tf-testacc6068554464636384429.xiaozhu.com
  block_ips.# = 0
  cache_config.# = 0
  cdn_type = web
  certificate_config.# = 1
  certificate_config.3358202545.private_key = -----BEGIN PRIVATE KEY-----
MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQDOdY9ITNHamP4B
3neS4bEm00oV5sNkqoNfGrwaR7qfwvGM76VXXb0VHnROeInXAn3QlYaFhLXDRVyJ
CqVg6SJhbaL6LqXTXQtl6ARxAuNqK/WQPsFbGgoofkkbgd6/sjc1x4fq5lGtBS3o
o/8dIFKL5s51LJO1DDM8CS2cJrsovCwrokyJUM0PUU3WmKNreDrhlZHUIic7GIkU
TTzqgirWL2HBAS2pSxTnFZ9GhvyPRZ9D963PHyvQRB2qGEbp5rrFAnQnCqYrBxym
HbZnd2YM/mfjwZIwEbbnzupDl0qL+ru2sy0MWYGSZnmoDZKkOP0fvW+BkOnK49D4
uueUFmVRAgMBAAECggEAGFrP2zSMsN/JXwkSS/Zpwm28WJcPR6nBs49gzyzU/BGw
EvMWKxc4vewIxlT71axKkTeCVe/QzUc6YkQqPCNkVd/sEN093JAmTxAurfIsR5MF
9c0hXBDXT+2NzDvmvfBVCPgPtYsT6Xgp8T6fUp1Ef5JrmnD2v62/wX5Hrhr3ixdp
FoEJ92/FJN+Dt/Duri481+RYqUUbNqiZAuEMsny/hM+aijtpQgZK8FqnVfjIOo0l
p4TwfDoxtTwkj0QeJxelX/sNSD+iQ5xVmnxQ9Q1qmKMk0M1Wl8318ovcrXDTN3RZ
R/Fp+lNw0aC4OLuOJDc7A13l5unHmex4Ka1kbjOZwQKBgQD8w0hJiVVmiojJ3/YX
06rG3ufGlKXu4qvRL6eXNqnDp8AnKXpjKiDV5lGz+vK0pcAyF5AhpddDN20Un+nR
4VU7NslNJdBfjQxfEKJRkGBP22C+pdlxOBDgdgouN9q7TTr9StaQC0KF+HUcF8fO
sYpEp1R+27iHie+F6uS78ur0mQKBgQDRGnbv8acTDTBW8nll1wTRa0+iVR2yxglS
ndC97HzYHOFb+caVmaM18QzHaaRvAhQ0bZXOjFV0QVv1nytlmHxsMlG/N6LDz5jA
NLGvbrzDzki+RUk0TWJuYpvdevsnwaCxMwyY6CQ5MKhKAvwgE8AhdAq5MGcE+6U/
oAOZNyOxeQKBgQCiCB2i5mLUpSIjJ2r+wzXK3sH9zvTAOpaiNsZcbTJOto67jB9k
ynDaLhdaJRjJLSgT9H700vc3o6RNgGXHoYedufU5e3AkkKrJlkQ3vTHAf4V5MaA+
sA5BlenYzv1s7IlQLlV1aYJvl2Kba7MukSlt8UZ9PCUC3i2pz3Zp9cMgoQKBgQCK
GkR7bMq/1nIausJa9IwGFC3gNP8MV6dInVqEVXCO+2QL7wetPm+A7NdXzPoBJwpZ
JhdO93ho89Hcg2eSDgf/HazH8eLaGH32U9cW2rhpShDZOcGDfaiI5y+yM8s1Erki
z2h+hLOH4g8D8ry6ItE+RvneHY2syNb3EqPNyZEVYQKBgGeC4KiyTpW+SqmkDRU/
/FxIAlZE2vWmtJptAUkmLEVoBi0/GvJoYYJR4Jhj4gS9tJcrI860X0j05NZHvIX8
UJ/vZZeiFRiatXYUtKbD/ngoOXTF92ew2J3eCwNHH5OKwEO3aejaYwD9m0QM4kTl
VLWIf79+G3NrxIeqPnqv+l3J
-----END PRIVATE KEY-----
  certificate_config.3358202545.server_certificate = -----BEGIN CERTIFICATE-----
MIICrDCCAZQCCQDApyXUTYDE+DANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA0q
LnhpYW96aHUuY29tMB4XDTE4MTIzMTEwMDY1OVoXDTE5MTIzMTEwMDY1OVowGDEW
MBQGA1UEAwwNKi54aWFvemh1LmNvbTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBAM51j0hM0dqY/gHed5LhsSbTShXmw2Sqg18avBpHup/C8YzvpVddvRUe
dE54idcCfdCVhoWEtcNFXIkKpWDpImFtovoupdNdC2XoBHEC42or9ZA+wVsaCih+
SRuB3r+yNzXHh+rmUa0FLeij/x0gUovmznUsk7UMMzwJLZwmuyi8LCuiTIlQzQ9R
TdaYo2t4OuGVkdQiJzsYiRRNPOqCKtYvYcEBLalLFOcVn0aG/I9Fn0P3rc8fK9BE
HaoYRunmusUCdCcKpisHHKYdtmd3Zgz+Z+PBkjARtufO6kOXSov6u7azLQxZgZJm
eagNkqQ4/R+9b4GQ6crj0Pi655QWZVECAwEAATANBgkqhkiG9w0BAQsFAAOCAQEA
JAWcQb7942jWDWjFqW5C4eyJBuJxK8dsTsYpOa0ccpp/A2cVCTNMzV/sMCGROy2B
uhrTMG4q3QawFwV+K4nQX8yJUJp57zyUfdit+yOD4hDzPyF5MOlOwzdqeg9Y9ODL
Sc2O6J3tFsx2332Hb7RXFlYodEk1uFezrj4YLVrAgk1QojaBHpuFiA/O3eCmErjW
c88bcil60qMvlhBhCWaZzivji6oc3y/6qWRZ3apxp9/6sjOvlm5Q+wwoLHXaU+L6
w7TZptw6upuSrAS+N4Rwqgn5TfPvbGkdRG0X0TDzE1+F+w377kE71nV8lmi8F786
bqloD1gWJER071WRz1coHQ==
-----END CERTIFICATE-----

  certificate_config.3358202545.server_certificate_status = on
  domain_name = tf-testacc6068554464636384429.xiaozhu.com
  http_header_config.# = 0
  optimize_enable = 
  page_compress_enable = 
  range_enable = 
  scope = overseas
  source_port = 80
  source_type = oss
  sources.# = 1
  sources.21942838 = terraformtest.aliyuncs.com
  video_seek_enable =
2019/01/04 11:38:55 [INFO] terraform: building graph: GraphTypeApply
2019/01/04 11:38:55 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
--- PASS: TestAccAlicloudCdnDomain_https (412.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	412.119s
```

Reply:
OK. Please check other comments.